src/main/java/mods/eln/misc/FloatLinear.kt

```diff
@@ -0,0 +1,130 @@
+    operator fun plus(other: Float) = add(other)
+    fun sub(other: Vec3f) = add(other.negate())
+    fun sub(other: Float) = add(-other)
+    operator fun minus(other: Vec3f) = sub(other)
+    operator fun minus(other: Float) = sub(other)
+    fun mul(other: Vec3f) = Vec3f(x * other.x, y * other.y, z * other.z)
+    fun mul(other: Float) = Vec3f(x * other, y * other, z * other)
+    operator fun times(other: Vec3f) = mul(other)
+    operator fun times(other: Float) = mul(other)
+    operator fun div(other: Vec3f) = mul(other.reciprocal())
+    operator fun div(other: Float) = mul(1/other)
+
+    fun magSquared() = dot(this)
+    fun mag() = Math.sqrt(magSquared().toDouble()).toFloat()
+    fun isZero() = magSquared() == 0.0f
+
+    fun distanceTo(other: Vec3f) = sub(other).mag()
+    fun normalized() = if(isZero()) Vec3f(0.0f, 0.0f, 0.0f) else div(mag())
+}
+
+data class Line3f(var origin: Vec3f, var vector: Vec3f) {
+    companion object {
+        fun fromEndpoints(a: Vec3f, b: Vec3f) = Line3f(a, b.sub(a))
+    }
+
+    fun pointAt(u: Float) = origin.add(vector.mul(u))
+    fun midpoint() = pointAt(0.5f)
+    fun opposite() = pointAt(1.0f)
+    fun length() = vector.mag()
+
+    fun allIntersectionsAlongU(): List<Float> {
+        var isctus = ArrayList<Float>()
+
+        for(axis in Vec3f.AXES) {
+            val kor = Math.floor(origin.dot(axis).toDouble()).toLong()
+            val kop = Math.floor(opposite().dot(axis).toDouble()).toLong()
+
+            for(k in (Math.min(kor, kop) - 1) .. (Math.max(kor, kop) + 1)) {
+                val u = Plane3f(axis, axis.mul(k.toFloat())).isctLineU(this)
+                if(u === null || u < 0.0 || u >= 1.0) continue
+                isctus.add(u)
+            }
+        }
+
+        isctus.sort()
+        return isctus
+    }
+    fun allIntersectionsAlong() = allIntersectionsAlongU().map {u -> pointAt(u)}
+
+    fun allSegmentsAlong(): List<Line3f> {
+        var segs = ArrayList<Line3f>()
+
+        val initSeq = ArrayList<Vec3f>()
+        initSeq.add(origin)
+        val iscts = allIntersectionsAlong()
+
+        for((a, b) in (initSeq.asSequence() + iscts.asSequence()).zip(iscts.asSequence() + opposite())) {
+            segs.add(Line3f(a, b.sub(a)))
+        }
+
+        return segs
+    }
+    fun allMidpointsAlong() = allSegmentsAlong().map {seg -> seg.midpoint()}
+    fun allCoordonatesAlong(dim: Int = 0) = allMidpointsAlong().map {pt -> pt.toCoordonate(dim)}
+}
+
+data class Plane3f(var normal: Vec3f, var offset: Float) {
+    init {
+        normal = normal.normalized()
+    }
+
+    companion object {
+        val PLANE_X = Plane3f(Vec3f.AXIS_X, 0.0f)
+        val PLANE_Y = Plane3f(Vec3f.AXIS_Y, 0.0f)
+        val PLANE_Z = Plane3f(Vec3f.AXIS_Z, 0.0f)
+        val AXIS_PLANES = arrayOf(PLANE_X, PLANE_Y, PLANE_Z)
+    }
+
+    constructor(normal: Vec3f, point: Vec3f) : this(normal, -normal.dot(point))
+
+    fun pointOn() = normal.mul(-offset)
+
+    fun isctLineU(ln: Line3f): Float? {
```

Comment:
What does "isctLine" do, exactly? Documentation is good, and better names would also be good.